Upgraded docker and Makefile

sphinxc0re opened the pull request:
## Summary

While I was moving the docker automated build to the `holochain` org, the build of our Dockerfile errored while installing clippy. It seems like clippy is now officially supported by the rust core team and inside the component collection of `rustup`. This means better support but also that I had to do some adjustments. As of this update we will be using:

 - `nightly-2018-07-17` for our tools (`rustfmt`, `clippy`)
 - `nightly-2018-06-01` for Wasm compilation
 - `stable` for everything else

The upgrade of `rustfmt` again will result in some reformatting. I intentionally didn't include the changes of that because the changes will still have to go though the docker hub build. Also as I suspected, the resulting reformatting won't be too heavy since it only touches some imports.

I also did some changes to the Makefile. I purposely removed the `--verbose` flag from the regular Rust build to (drum roll) reduce verbosity.

~Another thing I changed was the call to clippy. It now is called without any flags to prevent unwanted behavior. We should always stick to the default community standards and this helps a lot with doing so.~ 

*The CI will probably fail because it will be using the old docker image*

## What reviewers can do
You can check out the branch locally, run the build script (`. docker/build-ubuntu`), run the formatting script (`. docker/run-fmt`) and check whether the tests run using `. docker/run-test`

ddd-mtl commented:
> We should always stick to the default community standards and this helps a lot with doing so.
Again I strongly disagree with this.
Clippy has warnings for style things which do not impact code behavior.
Again, I want to be able to put 'return' when it makes the code easier to understand.
rustfmt and clippy has config files, we should take the time to agree on the configuration of those tools before enforcing them.

ddd-mtl commented:
I'm also getting tired of wasting time making my setup work because toolchain versions keep changing.
How likely are we going to stick to these new toolchain versions?

sphinxc0re commented:
@ddd-mtl You shouldn't have to worry about toolchain versions if you are using the docker setup locally. That's what the docker setup is for. You shouldn't have to install those tools locally and maintain sync with the project manually.
But to answer your question: I don't know how likely it is that we stick to those versions. For now I think it is at least more likely than that we change them every month. I pinned every version on purpose.
Clippy is only throwing errors ATM so it shouldn't affect the code.

ddd-mtl commented:
@sphinxc0re I don't want to use docker on windows, but I am willing to give it a try if it actually works and makes that pain go away. Maybe Ashanti could help me out with this, as they are also on windows.

zippy commented:
FYI the new run of `. docker/run-test` still saves built files as root:

```
ll /home/eric/code/metacurrency/holochain/holochain-rust/core/src/nucleus/wasm-test/target/
total 20
drwxr-xr-x 4 root root 4096 Jul 18 11:43 ./
drwxrwxr-x 4 eric eric 4096 Jul 18 11:43 ../
drwxr-xr-x 8 root root 4096 Jul 18 11:43 debug/
-rw-r--r-- 1 root root 1521 Jul 18 11:43 .rustc_info.json
drwxr-xr-x 3 root root 4096 Jul 18 11:43 wasm32-unknown-unknown/
eric@isotopy:~/code/metacurrency/holochain/holochain-rust$ 
```

thedavidmeister commented:
@sphinxc0re this thread shows why we shouldn't try to change more than one global config per branch, the discussion around clippy config changes are now derailing the point of the branch which is a simple version bump... besides, there's already a separate ticket up to deal with clippy warnings at #105 

as a general point, it seems to me that sooner or later we will need a clear schedule to bump versions and that the frequency of this shouldn't be determined by BYO core dev environments. we're building _primarily_ for an ecosystem of apps, hackathons, etc. not an internal product.

for this branch, can we please do the bare minimum to get clippy working on the new version, with the same config we had before and then we can just approve the version itself?

@ddd-mtl for the same reason, i don't think this is the appropriate place to air complaints about docker itself, this branch should just focus on bumping nightly, it's been ~1.5 months since the last version so we probably are due for an update nowish anyway, right?

whatever we do locally, there has to be some deterministic or near-deterministic CI environment or CI simply breaks and/or burns more time than it saves.

you don't have to run docker locally for dev (i personally use it very little) but then you do have to maintain toolchains, i don't know of any alternative unfortunately :/

the only real requirements/guarantees are:

- CI must pass or fail reliably for any given commit
- CI must pass before merging a PR
- we maintain some shared debugging environment(s) somewhere/somehow that reliably passes/fails builds in the same way as CI

docker is not the only way to achieve this (for example we could use vagrant to build our environment), but it is the most popular and best supported option by the various hosted CI providers.

@ddd-mtl if you try out docker on windows and run into problems, i'm happy to help where i can.

ddd-mtl commented:
@thedavidmeister I don't get why we should update versions every 1.5 month. Seems obvious to you, but it is not for me.  Would you mind explaining me?

thedavidmeister commented:
@ddd-mtl i didn't say we'd do it every 1.5 months, not sure where you got that from

thedavidmeister commented:
i opened a ticket to discuss clippy usage https://github.com/holochain/holochain-rust/issues/151

sphinxc0re commented:
@zippy @thedavidmeister @ddd-mtl ping!

sphinxc0re commented:
This needs to get merges ASAP. So please take another look at this

zippy commented:
looking much better, now using correct UID for me instead of root.

I see a few more problems:

- I think it would be good to actually commit the changes made by `. docker/run-fmt` to this branch so that I can confirm that run-fmt is doing what you expect it to when I run it, i.e. not make any changes.  Also if we don't I think the develop branch will show up broken for CI as soon as this gets merged.

- Every time I run `. docker/run-test` it seems seems to be re-downloading and re-compiling everything.  Is that expected?  I would have expected some caching.  I also see the same library (e.g. `serde_derive v1.0.70`) being recompiled multiple times per run.  Is that expected?

- c_binding_tests failing

Below is an example run:

```
$ . docker/run-test 
cargo  +nightly-2018-07-17 fmt -- --check
cargo  +nightly-2018-07-17 clippy -- -A needless_return
 Downloading serde_json v1.0.22
 Downloading base64 v0.9.2                                                                                                                                                                                          
 Downloading proc-macro2 v0.4.6                                                                                                                                                                                     
 Downloading indexmap v1.0.1                                                                                                                                                                                        
 Downloading safemem v0.2.0                                                                                                                                                                                         
 Downloading chrono v0.4.4                                                                                                                                                                                          
 Downloading rust-base58 v0.0.4                                                                                                                                                                                     
 Downloading wasmi v0.3.0                                                                                                                                                                                           
 Downloading multihash v0.8.0                                                                                                                                                                                       
 Downloading snowflake v1.3.0                                                                                                                                                                                       
 Downloading num-integer v0.1.39                                                                                                                                                                                    
 Downloading time v0.1.40                                                                                                                                                                                           
 Downloading num v0.2.0                                                                                                                                                                                             
 Downloading num-bigint v0.2.0                                                                                                                                                                                      
 Downloading num-iter v0.1.37                                                                                                                                                                                       
 Downloading num-complex v0.2.0                                                                                                                                                                                     
 Downloading num-rational v0.2.1                                                                                                                                                                                    
 Downloading nan-preserving-float v0.1.0                                                                                                                                                                            
 Downloading memory_units v0.3.0                                                                                                                                                                                    
 Downloading parity-wasm v0.31.0                                                                                                                                                                                    
 Downloading sha2 v0.7.1                                                                                                                                                                                            
 Downloading tiny-keccak v1.4.2                                                                                                                                                                                     
 Downloading sha1 v0.5.0                                                                                                                                                                                            
 Downloading byte-tools v0.2.0                                                                                                                                                                                      
 Downloading digest v0.7.4                                                                                                                                                                                          
 Downloading fake-simd v0.1.2                                                                                                                                                                                       
 Downloading block-buffer v0.3.3                                                                                                                                                                                    
 Downloading generic-array v0.9.0                                                                                                                                                                                   
 Downloading typenum v1.10.0                                                                                                                                                                                        
 Downloading arrayref v0.3.4                                                                                                                                                                                        
 Downloading crunchy v0.1.6                                                                                                                                                                                         
 Downloading wabt v0.2.2                                                                                                                                                                                            
 Downloading wabt-sys v0.1.4                                                                                                                                                                                        
 Downloading cc v1.0.17                                                                                                                                                                                             
    Finished dev [unoptimized + debuginfo] target(s) in 7.03s                                                                                                                                                       
cargo  build --all
   Compiling num-traits v0.2.5
   Compiling typenum v1.10.0
   Compiling num-integer v0.1.39
   Compiling num-bigint v0.2.0
   Compiling unicode-xid v0.1.0
   Compiling crunchy v0.1.6
   Compiling num-iter v0.1.37
   Compiling num-complex v0.2.0
   Compiling num-rational v0.2.1
   Compiling serde v1.0.70
   Compiling libc v0.2.42
   Compiling arrayref v0.3.4
   Compiling byte-tools v0.2.0
   Compiling cc v1.0.17
   Compiling indexmap v1.0.1
   Compiling cfg-if v0.1.4
   Compiling byteorder v1.2.3
   Compiling safemem v0.2.0
   Compiling dtoa v0.4.3
   Compiling fake-simd v0.1.2
   Compiling itoa v0.4.2
   Compiling memory_units v0.3.0
   Compiling sha1 v0.5.0
   Compiling nan-preserving-float v0.1.0
   Compiling remove_dir_all v0.5.1
   Compiling holochain_agent v0.1.0 (file:///holochain/agent)
   Compiling snowflake v1.3.0
   Compiling proc-macro2 v0.4.6
   Compiling block-buffer v0.3.3
   Compiling rand v0.4.2
   Compiling time v0.1.40
   Compiling parity-wasm v0.31.0
   Compiling base64 v0.9.2
   Compiling cmake v0.1.31
   Compiling uuid v0.6.5
   Compiling tempdir v0.3.7
   Compiling quote v0.6.3
   Compiling tiny-keccak v1.4.2
   Compiling wabt-sys v0.1.4
   Compiling generic-array v0.9.0
   Compiling wasmi v0.3.0
   Compiling syn v0.14.4
   Compiling digest v0.7.4
   Compiling chrono v0.4.4
   Compiling serde_json v1.0.22
   Compiling serde_derive v1.0.70
   Compiling sha2 v0.7.1
   Compiling multihash v0.8.0
   Compiling num v0.2.0
   Compiling rust-base58 v0.0.4
   Compiling holochain_dna v0.1.0 (file:///holochain/dna)
   Compiling wabt v0.2.2
   Compiling holochain_core v0.1.0 (file:///holochain/core)
   Compiling holochain_dna_c_binding v0.1.0 (file:///holochain/dna_c_binding)
   Compiling holochain_core_api v0.1.0 (file:///holochain/core_api)
   Compiling test_utils v0.1.0 (file:///holochain/test_utils)
   Compiling holochain_test_bin v0.1.0 (file:///holochain/test_bin)
   Compiling holochain_core_api_c_binding v0.1.0 (file:///holochain/core_api_c_binding)
    Finished dev [unoptimized + debuginfo] target(s) in 47.78s
cd core/src/nucleus/wasm-test && cargo  +nightly-2018-06-01 build --target wasm32-unknown-unknown
    Finished dev [unoptimized + debuginfo] target(s) in 0.08s
cd core_api/wasm-test/round_trip && cargo  +nightly-2018-06-01 build --target wasm32-unknown-unknown
 Downloading serde_derive v1.0.66
 Downloading serde v1.0.66                                                      
 Downloading syn v0.14.2                                                        
 Downloading itoa v0.4.1                                                        
 Downloading dtoa v0.4.2                                                        
    Finished dev [unoptimized + debuginfo] target(s) in 1.71s                   
cd core_api/wasm-test/commit && cargo  +nightly-2018-06-01 build --target wasm32-unknown-unknown
    Finished dev [unoptimized + debuginfo] target(s) in 0.12s
RUSTFLAGS="-D warnings" cargo  test
 Downloading wabt v0.4.0
 Downloading wabt-sys v0.2.0                                                    
   Compiling num-traits v0.2.5                                                  
   Compiling typenum v1.10.0
   Compiling num-integer v0.1.39
   Compiling unicode-xid v0.1.0
   Compiling num-bigint v0.2.0
   Compiling serde v1.0.70
   Compiling num-rational v0.2.1
   Compiling cc v1.0.17
   Compiling crunchy v0.1.6
   Compiling num-iter v0.1.37
   Compiling num-complex v0.2.0
   Compiling byte-tools v0.2.0
   Compiling libc v0.2.42
   Compiling arrayref v0.3.4
   Compiling cfg-if v0.1.4
   Compiling itoa v0.4.2
   Compiling dtoa v0.4.3
   Compiling safemem v0.2.0
   Compiling indexmap v1.0.1
   Compiling byteorder v1.2.3
   Compiling fake-simd v0.1.2
   Compiling remove_dir_all v0.5.1
   Compiling memory_units v0.3.0
   Compiling nan-preserving-float v0.1.0
   Compiling sha1 v0.5.0
   Compiling holochain_agent v0.1.0 (file:///holochain/agent)
   Compiling snowflake v1.3.0
   Compiling proc-macro2 v0.4.6
   Compiling block-buffer v0.3.3
   Compiling cmake v0.1.31
   Compiling rand v0.4.2
   Compiling time v0.1.40
   Compiling parity-wasm v0.31.0
   Compiling base64 v0.9.2
   Compiling wabt-sys v0.1.4
   Compiling wabt-sys v0.2.0
   Compiling quote v0.6.3
   Compiling uuid v0.6.5
   Compiling tempdir v0.3.7
   Compiling generic-array v0.9.0
   Compiling tiny-keccak v1.4.2
   Compiling syn v0.14.4
   Compiling wasmi v0.3.0
   Compiling digest v0.7.4
   Compiling chrono v0.4.4
   Compiling serde_json v1.0.22
   Compiling serde_derive v1.0.70
   Compiling sha2 v0.7.1
   Compiling multihash v0.8.0
   Compiling num v0.2.0
   Compiling rust-base58 v0.0.4
   Compiling holochain_dna v0.1.0 (file:///holochain/dna)
   Compiling wabt v0.2.2
   Compiling wabt v0.4.0
   Compiling holochain_core v0.1.0 (file:///holochain/core)
   Compiling holochain_dna_c_binding v0.1.0 (file:///holochain/dna_c_binding)
   Compiling test_utils v0.1.0 (file:///holochain/test_utils)
   Compiling holochain_core_api v0.1.0 (file:///holochain/core_api)
   Compiling holochain_test_bin v0.1.0 (file:///holochain/test_bin)
   Compiling holochain_core_api_c_binding v0.1.0 (file:///holochain/core_api_c_binding)
    Finished dev [unoptimized + debuginfo] target(s) in 1m 2.08s
     Running target/debug/deps/holochain_agent-6fb08c68abc13661

running 1 test
test tests::can_instantiate_agent ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

     Running target/debug/deps/holochain_core-8e7f3b7430ac8da4

running 59 tests
test agent::keys::tests::key_new ... ok
test chain::entry::tests::content ... ok
test chain::entry::tests::eq ... ok
test chain::entry::tests::entry_type ... ok
test chain::entry::tests::hash_entry_type ... ok
test agent::keys::tests::keys_new ... ok
test chain::entry::tests::hash_known ... ok
test chain::entry::tests::new ... ok
test chain::entry::tests::hash_content ... ok
test chain::entry::tests::validate ... ok
test chain::header::tests::entry ... ok
test chain::header::tests::entry_type ... ok
test chain::header::tests::hash_entry_type ... ok
test chain::header::tests::hash_known ... ok
test chain::header::tests::hash_type_next ... ok
test chain::header::tests::new ... ok
test chain::header::tests::hash_entry_content ... ok
test chain::header::tests::hash_chain_state ... ok
test chain::header::tests::eq ... ok
test chain::header::tests::signature ... ok
test chain::header::tests::validate ... ok
test chain::header::tests::next ... ok
test chain::header::tests::time ... ok
test chain::header::tests::type_next ... ok
test chain::memory::tests::iter ... ok
test chain::memory::tests::get ... ok
test chain::memory::tests::new ... ok
test chain::memory::tests::json_round_trip ... ok
test chain::memory::tests::into_iter ... ok
test chain::memory::tests::get_entry ... ok
test chain::memory::tests::push ... ok
test chain::pair::tests::entry ... ok
test chain::pair::tests::header ... ok
test chain::memory::tests::top ... ok
test chain::pair::tests::validate ... ok
test error::tests::can_instantiate ... ok
test error::tests::can_raise_holochain_error ... ok
test error::tests::can_return_result ... ok
test hash::tests::bytes_to_b58_known_golang ... ok
test hash::tests::str_to_b58_hash_known_golang ... ok
test chain::pair::tests::new ... ok
test nucleus::tests::can_instantiate_nucleus_state ... ok
test nucleus::tests::can_reduce_execfn_action ... ok
test chain::memory::tests::top_type ... ok
test nucleus::ribosome::tests::test_print ... ok
test nucleus::tests::can_reduce_initialize_action ... ok
test chain::memory::tests::validate ... ok
test persister::tests::can_instantiate ... ok
test persister::tests::can_roundtrip ... ok
test nucleus::tests::can_reduce_return_init_result_action ... ok
test tests::call_ribosome_wrong_dna ... ok
test tests::call_ribosome_wrong_function ... ok
test tests::can_dispatch_with_observer ... ok
test tests::call_wrong_ribosome_function ... ok
test tests::can_dispatch_and_wait ... ok
test tests::call_ribosome_function ... ok
test tests::test_genesis_err ... ok
test tests::test_missing_genesis ... ok
test tests::test_genesis_ok ... ok

test result: ok. 59 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

     Running target/debug/deps/holochain_core_api-f5f3e8e9bfcfd324

running 8 tests
test tests::can_get_state ... ok
test tests::can_instantiate ... ok
test tests::can_start_and_stop ... ok
test tests::can_call ... ok
test tests::fails_instantiate_if_genesis_fails ... ok
test tests::can_call_test ... ok
test tests::can_call_commit ... ok
test tests::fails_instantiate_if_genesis_times_out ... ok

test result: ok. 8 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

     Running target/debug/deps/holochain_core_api_c_binding-4fc26192741d4a03

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

     Running target/debug/deps/holochain_dna-5236037f45b542aa

running 16 tests
test tests::get_wasm_for_capability ... ok
test tests::can_parse_and_output_json_helpers ... ok
test tests::can_parse_and_output_json ... ok
test tests::default_value_test ... ok
test tests::parse_fail_if_bad_type_dna ... ok
test tests::parse_fail_if_bad_type_entry_type ... ok
test tests::parse_accepts_arbitrary_dna_properties ... ok
test tests::parse_wasm ... ok
test tests::parse_fail_if_bad_type_zome ... ok
test tests::parse_with_defaults_dna ... ok
test tests::parse_with_defaults_entry_type ... ok
test zome::capabilities::tests::build_and_compare ... ok
test tests::parse_with_defaults_zome ... ok
test zome::tests::build_and_compare ... ok
test zome::entry_types::tests::build_and_compare ... ok
test tests::parse_and_serialize_compare ... ok

test result: ok. 16 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

     Running target/debug/deps/holochain_dna_c_binding-110bd3af2e4de67c

running 2 tests
test tests::can_set_and_get_value ... ok
test tests::serialize_and_deserialize ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

     Running target/debug/deps/holochain_test_bin-244779809a740b97

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

     Running target/debug/deps/test_utils-0b3e0e24cd9152da

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

   Doc-tests holochain_agent

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

   Doc-tests holochain_core_api

running 1 test
test src/lib.rs -  (line 5) ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

   Doc-tests holochain_core

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

   Doc-tests holochain_dna

running 5 tests
test src/lib.rs - Dna::to_json (line 129) ... ok
test src/lib.rs - Dna::new (line 97) ... ok
test src/lib.rs -  (line 8) ... ok
test src/lib.rs - Dna::new_from_json (line 112) ... ok
test src/lib.rs - Dna::to_json_pretty (line 144) ... ok

test result: ok. 5 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

   Doc-tests test_utils

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

qmake -o c_binding_tests/hc_dna/Makefile c_binding_tests/hc_dna/qmake.pro
Failure to open file: /holochain/c_binding_tests/hc_dna/Makefile
Unable to generate makefile for: c_binding_tests/hc_dna/qmake.pro
Makefile:40: recipe for target 'c_binding_tests/hc_dna/' failed
make: *** [c_binding_tests/hc_dna/] Error 5
```

sphinxc0re commented:
@zippy caching crates from the docker side is a problem I'm trying to solve as well. But I think this is a separate issue.

sphinxc0re commented:
Hmm... test are working for me.

sphinxc0re commented:
@zippy try deleting the `Makefile` and the `.qmake.stash` in the c bindings test dir.

zippy commented:
ok, worked after deleting Makefile and .stash.  Should we be adding that into script some place?

sphinxc0re commented:
@zippy no, adding this to a script just produces unnecessary overhead. I believe this was just a one-off thing

sphinxc0re commented:
@zippy Also, I added crate caching. Please take one last look at the code I added. The Docker container mounts your local `.cargo/registry` directory that is responsible for caching crates.

sphinxc0re commented:
The build won't work, because we are using a non-existent image
